File: Python/tigre/utilities/visualization/animate_geometry.py

import numpy as np
import copy
from matplotlib import pyplot as plt
from scipy.spatial.transform import Rotation
import mpl_toolkits.mplot3d.art3d as art3d
from matplotlib import animation
import tigre

   
def animate_geometry(ogeo,angles=None,pos=None,rotation="SD",animate=True,fname=None):
    """
    Plot animated TIGRE geometry

    Parameters
    ----------
    ogeo : tigre geometry object
        Geometry to be plotted.
    angles : array, optional
        Angles to plot for the geometry. The default is np.linspace(0,2*np.pi,100).
    pos : float, optional
        A specific angles[pos] for static geometry plot. The default is 0, 
        angles[0] is used.
    rotation : string, optional
        Style of animated plot with rotation style: "SD" - fixed object, source 
        and detector moving; "obj" - source and detector fixed, object moves, 
        only applicable when source and detector are not changing relative positions. 
        The default is "obj".
    animate : bool, optional
        Plot animated geometry setup or static geometry. The default is True.
    fname : string, optional
        Filename to save animated geometry plotting. The default is None.

    Returns
    -------
    Handle for animation or plot.

    """
#PLOT_GEOMETRY(GEO,ANGLES,POS,Animate,fname) plots a simplified version of the CBCT geometry with the
# given geomerty GEO and scanning angles ANGLES at angle POS. If angles is 
# not given, [0,2pi] at 100 steps will be given, and pos=0 will be chosen. If 
# Animate=TRUE, an animation file will be generated. Animation file name can be
# specified by fname.
# 
# h=PLOT_GEOMETRY(...) will return the figure handle 
#--------------------------------------------------------------------------
#--------------------------------------------------------------------------
# This file is part of the TIGRE Toolbox
# 
# Copyright (c) 2015, University of Bath and 
#                     CERN-European Organization for Nuclear Research
#                     All rights reserved.
#
# License:            Open Source under BSD. 
#                     See the full license at
#                     https://github.com/CERN/TIGRE/blob/master/LICENSE
#
# Contact:            tigre.toolbox@gmail.com
# Codes:              https://github.com/CERN/TIGRE/
# Coded by:           Ander Biguri, modified by Yi Liu
#--------------------------------------------------------------------------

    # check geo, which makes DSD, DSO etc all matrices
    if angles is None:
        angles = ogeo.angles if hasattr(ogeo,"angles") else np.linspace(0,2*np.pi,100)
        
    geo = copy.deepcopy(ogeo)
    try:
        geo.check_geo(angles)
    except:
        raise ValueError("Geometry and angles are inconsistent. Please check inputs.")
        
    # Validate pos index
    pos = 0 if pos is None or pos < 0 or pos >= len(angles) else pos
            
    # Set up figure 
    fig=plt.figure(figsize=(10,8))
    ax = fig.add_subplot(111, projection='3d')
    ax.view_init(azim=52,elev=26)       
    ln = geo.sVoxel.min()/2   

    # Define rotation matrices
    Rs = Rotation.from_euler('ZYZ', geo.angles).as_matrix()
    
    # Compute source and detector centres
    scent = np.column_stack((geo.DSO, np.zeros_like(geo.angles[:, 1]), np.zeros_like(geo.angles[:, 2])))
    dcent = np.column_stack((-geo.DSD+geo.DSO, geo.offDetector[:,1], geo.offDetector[:,0]))
   
    def plot_trajectory(points, color, label=None, marker_size=2.5):
        ax.plot3D(points[:, 0], points[:, 1], points[:, 2], color=color, ls='', marker=".", markersize=marker_size, mfc=color, mec=color)
        if label == None:
            return ax.scatter(points[pos, 0], points[pos, 1], points[pos, 2], color=color, s=5)
        else:
            return ax.scatter(points[pos, 0], points[pos, 1], points[pos, 2], color=color, s=5), ax.text(points[pos, 0], points[pos, 1], points[pos, 2] + 30, label)

    # Compute source trajectory, coordinates in (z,y,x)
    if rotation == "SD":
        stj = np.einsum('ijk,ik->ij', Rs, scent)
        if hasattr(geo, 'COR'):
            stj[:, 1] += geo.COR.flatten()
    elif rotation == "obj":
        stj = np.zeros_like(geo.angles)
        stj[pos,:] = np.matmul(Rs[pos,:,:], scent[pos,:])
        if hasattr(geo, 'COR'):
            stj[pos,1] += geo.COR[pos]
    
    source, stext = plot_trajectory(stj, 'plum', 'S')
        
    # Compute detector trajectory
    if rotation == "SD":
        dtj = np.einsum('ijk,ik->ij', Rs, dcent)
        if hasattr(geo, 'COR'):
            dtj[:, 1] += geo.COR.flatten()
    elif rotation == "obj":
        dtj = np.zeros_like(geo.angles)
        dtj[pos,:] = np.matmul(Rs[pos,:,:], dcent[pos,:]) 
        if hasattr(geo, 'COR'):
            dtj[:,1] += geo.COR.flatten()

    det, dtext = plot_trajectory(dtj, 'skyblue', 'D')
                
    # Compute detector orientation
    R_detector = np.matmul(Rs, Rotation.from_euler('XYZ', geo.rotDetector).as_matrix())
    
    # Detector cube, cp returns four cordinates of corners closest to source
    ddp = 30 # detector depth
    dsz = np.array([ddp, geo.sDetector[1], geo.sDetector[0]])  # at angles (0,0,0)
    dverts = calCube(dtj, dsz, R_detector, offcent=np.array([-ddp / 2, 0, 0]))
    dcube = art3d.Poly3DCollection(dverts[pos], color='brown', alpha=0.2)
    ax.add_collection3d(dcube)
        
    
    # Compute image origin trajectory. NOTE: offOrigin is in (z,y,x)
    otj = np.fliplr(geo.offOrigin).astype(np.float32)
    # # displacement in y for geo.COR
    if hasattr(geo, 'COR'):
        otj[:,1] += geo.COR.flatten()   
    _ = plot_trajectory(otj, 'grey')    
    
    # Compute object cube
    overts = calCube(otj, geo.sVoxel[[2, 1, 0]], np.swapaxes(Rs, 1, 2) if rotation == "obj" else 'eye')
    ocube = art3d.Poly3DCollection(overts[pos], color='c', alpha=0.2)
    ax.add_collection3d(ocube)    
    
    # Cordinates Arrows from origin
    ax.quiver(0, 0, 0, 1, 0, 0, length=ln, color='r')
    ax.quiver(0, 0, 0, 0, 1, 0, length=ln, color='b')
    ax.quiver(0, 0, 0, 0, 0, 1, length=ln, color='g')
    ax.text(-10, -10, -10, 'O', None)
        
    # Effective beam profile
    beampf = [ax.plot3D(*zip(stj[pos, :], dverts[pos][0][i]), color='y') for i in range(4)]

    # Cenrtal beam
    cbeam=ax.plot3D(*zip(stj[pos,:],dtj[pos,:]),color='pink')
    
    # set tight limits and aspect
    lims = [ax.get_xlim3d(), ax.get_ylim3d(), ax.get_zlim3d()]
    ax.set_box_aspect([lim[1] - lim[0] for lim in lims])

    # Set up plot title and axis label
    roll, pitch, yaw = geo.rotDetector[:,0]/np.pi*180, geo.rotDetector[:,1]/np.pi*180, geo.rotDetector[:,2]/np.pi*180
    if rotation == 'SD':
        ang = geo.angles[pos] / np.pi * 180
        title = ax.set_title(
            f"CBCT geometry in scale - Source at angle [{ang[0]:.1f}, {ang[1]:.1f}, {ang[2]:.1f}]°\n\n"
            f"Detector rotation [{roll[pos]:.1f}, {pitch[pos]:.1f}, {yaw[pos]:.1f}]°\n\n"
            f"Detector offset [{geo.offDetector[pos, 0]:.1f}, {geo.offDetector[pos, 1]:.1f}] mm")
    elif rotation == 'obj':
        ang = -geo.angles[pos] / np.pi * 180
        title = ax.set_title(
            f"CBCT geometry in scale - Object at angle [{ang[0]:.1f}, {ang[1]:.1f}, {ang[2]:.1f}]°\n\n"
            f"Detector rotation [{roll[pos]:.1f}, {pitch[pos]:.1f}, {yaw[pos]:.1f}]°\n\n"
            f"Detector offset [{geo.offDetector[pos, 0]:.1f}, {geo.offDetector[pos, 1]:.1f}] mm")
    ax.set_xlabel('X');
    ax.set_ylabel('Y');
    ax.set_zlabel('Z');
          
    # Define update function
    def update(pos):
        ocube.set_verts(overts[pos]) # to move or rotate
        if rotation == 'SD':
            source.set_offsets(stj[pos, :2])
            source.set_3d_properties(stj[pos, 2], 'z')
            stext.set_position(stj[pos,:2] + 15)
            det.set_offsets(dtj[pos, :2])
            det.set_3d_properties(dtj[pos, 2], 'z')
            dtext.set_position(dtj[pos, :2] + 15)
            cbeam[0].set_data(*zip(stj[pos, :2], dtj[pos, :2]))
            cbeam[0].set_3d_properties((stj[pos, 2], dtj[pos, 2]),'z')
            for i in range(4):
                beampf[i][0].set_data(*zip(stj[pos, :2],dverts[pos][0][i][:2]))
                beampf[i][0].set_3d_properties((stj[pos, 2], dverts[pos][0][i][2]), 'z')
            dcube.set_verts(dverts[pos])
            ang = geo.angles[pos] / np.pi * 180
            title.set_text(
                f"CBCT geometry in scale - Source at angle [{ang[0]:.1f}, {ang[1]:.1f}, {ang[2]:.1f}]°\n\n"
                f"Detector rotation [{roll[pos]:.1f}, {pitch[pos]:.1f}, {yaw[pos]:.1f}]°\n\n"
                f"Detector offset [{geo.offDetector[pos, 0]:.1f}, {geo.offDetector[pos, 1]:.1f}] mm")
        elif rotation == "obj":
            ang = -geo.angles[pos] / np.pi * 180
            ax.set_title(
                f"CBCT geometry in scale - Object at angle [{ang[0]:.1f}, {ang[1]:.1f}, {ang[2]:.1f}]°\n\n"
                f"Detector rotation [{roll[pos]:.1f}, {pitch[pos]:.1f}, {yaw[pos]:.1f}]°\n\n"
                f"Detector offset [{geo.offDetector[pos, 0]:.1f}, {geo.offDetector[pos, 1]:.1f}] mm")
    
    if animate:
        ani = animation.FuncAnimation(fig, update, len(angles), interval=100)
        if fname:
            fname += '_geometry'
            for writer in ['ffmpeg','pillow','imagemagick']:
                try:
                    ani.save(f'{fname}.mp4' if writer == 'ffmpeg' else f'{fname}.gif', writer=writer, fps=30)
                    break
                except ValueError:
                    continue
        return ani
    else:
        return ax


def calCube(centre, size, R='eye', offcent=None):
    """
    Calculate vertices of a cuboid centered at 'centre' with given 'size' and rotation 'R'.
    Args:
        centre: array-like, shape (3,) or (n,3) - center position(s)
        size: float - cuboid size
        R: str or array-like, shape (3,3) or (n,3,3) - rotation matrix, 'eye' for identity
        offcent: array-like, shape (3,) - offset from center, defaults to [0,0,0]
    Returns:
        list of vertex arrays representing cuboid faces
    """
    # Define cube corners once as a constant
    CORNERS = np.array([[-1, -1, -1],
                        [ 1, -1, -1],
                        [ 1,  1, -1],
                        [-1,  1, -1],
                        [-1, -1,  1],
                        [ 1, -1,  1],
                        [ 1,  1,  1],
                        [-1,  1,  1]], dtype=np.float64)
    
    # Define face indices once as a constant
    FACE_INDICES = np.array([[1, 2, 6, 5],  # front
                            [0, 1, 2, 3],   # bottom
                            [4, 5, 6, 7],   # top
                            [0, 1, 5, 4],   # left
                            [2, 3, 7, 6],   # right
                            [0, 3, 7, 4]])  # back
    
    # Convert inputs to numpy arrays and set defaults
    centre = np.asarray(centre, dtype=np.float64)
    offcent = np.zeros(3, dtype=np.float64) if offcent is None else np.asarray(offcent, dtype=np.float64)
    is_batched = centre.ndim > 1
    
    # Pre-compute scaled corners
    scaled_corners = CORNERS * (size / 2) + offcent
    
    # Compute vertices
    if is_batched:
        n = centre.shape[0]
        verts = []
        # Handle rotation
        if isinstance(R, str):
            rotation = np.eye(3)
            rotated = scaled_corners @ rotation.T
            for i in range(n):
                corners = rotated + centre[i]
                verts.append(corners[FACE_INDICES])
        else:
            R = np.asarray(R)
            for i in range(n):
                corners = scaled_corners @ R[i].T + centre[i]
                verts.append(corners[FACE_INDICES])
    else:
        # Handle single cuboid
        if isinstance(R, str):
            rotation = np.eye(3)
        else:
            rotation = np.asarray(R)
        corners = scaled_corners @ rotation.T + centre
        verts = corners[FACE_INDICES].tolist()
    return verts


# self testing 
if __name__ == "__main__":
    n_voxel_z = 128
    n_voxel_y = 256
    n_voxel_x = 512
    n_detector_u = 400
    n_detector_v = 300
    off_detector_u = 0  # 500
    off_detector_v = 0  # 500
    off_origin_x = 0  # 300
    off_origin_y = 0  # 100
    off_origin_z = 0  # 100
    rot_detector_roll = 0 # degrees
    rot_detector_pitch = 0
    rot_detector_yaw = 0

    geo = tigre.geometry(mode="cone", default=True)
    geo.nVoxel = np.array([n_voxel_z, n_voxel_y, n_voxel_x])
    geo.sVoxel = geo.nVoxel
    geo.dVoxel = geo.sVoxel / geo.nVoxel
    geo.nDetector = np.array([n_detector_v, n_detector_u])
    geo.sDetector = geo.nDetector * geo.dDetector
    geo.offDetector = np.array([off_detector_v, off_detector_u])
    geo.offOrigin = np.array([off_origin_z, off_origin_y, off_origin_x])
    geo.rotDetector = np.radians([rot_detector_roll, rot_detector_pitch, rot_detector_yaw]) 
    print(geo)
    angles = np.linspace(0.0, 2*np.pi, 100)
    pos = np.argmin(abs(angles < np.pi / 6))
    ani1=animate_geometry(geo, angles, pos, rotation="obj", animate=True, fname="Rotate object")
    ani1
    plt.show()
    ani2=animate_geometry(geo, angles, pos, rotation="SD", animate=True, fname="Rotate source and detector")
    ani2
    plt.show()

File: MATLAB/Utilities/animateGeometry.m
function h = animateGeometry(geo, angles, pos, rotation, animate, fname)
%ANIMATEGEOMETRY Plot an animated (or static) TIGRE cone-beam geometry.
%
%   H = ANIMATEGEOMETRY(GEO, ANGLES, POS, ROTATION, ANIMATE, FNAME) draws the
%   source, detector (with its rotation rotDetector), object cube (with its
%   offset offOrigin / COR) and beam profile IN SCALE for geometry GEO at
%   the projection ANGLES, and steps through every view.
%
%   ANGLES   1xN or 3xN (ZYZ Euler, as Ax accepts). Default linspace(0,2*pi,100).
%   POS      index of the view drawn in a static plot / first frame (1-based).
%            Default 1.
%   ROTATION 'SD'  - object fixed, source and detector move (default);
%            'obj' - source and detector fixed, object rotates (only
%                    meaningful when source and detector do not move
%                    relative to each other).
%   ANIMATE  true (default) steps through all views; false draws view POS.
%   FNAME    if given, the animation is written to [FNAME '_geometry.gif'].
%
%   Returns the figure handle.
%
%   MATLAB port of tigre.utilities.visualization.animate_geometry. Uses
%   only base-MATLAB graphics (plot3/patch/quiver3/imwrite), no toolboxes.
%--------------------------------------------------------------------------
%--------------------------------------------------------------------------
% This file is part of the TIGRE Toolbox
%
% Copyright (c) 2015, University of Bath and
%                     CERN-European Organization for Nuclear Research
%                     All rights reserved.
%
% License:            Open Source under BSD.
%                     See the full license at
%                     https://github.com/CERN/TIGRE/blob/master/LICENSE
%
% Contact:            tigre.toolbox@gmail.com
% Codes:              https://github.com/CERN/TIGRE/
% Coded by:           Ander Biguri (plotgeometry), animated port by Yi Liu
%--------------------------------------------------------------------------

%% Arguments
if nargin < 2 || isempty(angles)
    angles = linspace(0, 2*pi, 100);
end
if nargin < 3 || isempty(pos), pos = 1; end
if nargin < 4 || isempty(rotation), rotation = 'SD'; end
if nargin < 5 || isempty(animate), animate = true; end
if nargin < 6, fname = ''; end
assert(any(strcmp(rotation, {'SD', 'obj'})), ...
    'TIGRE:animateGeometry:rotation', 'rotation must be ''SD'' or ''obj''');

angles = double(angles);
if size(angles, 1) == 1
    angles = [angles; zeros(2, size(angles, 2))];
end
assert(size(angles, 1) == 3, 'TIGRE:animateGeometry:angles', ...
    'angles must be 1xN or 3xN');
n = size(angles, 2);
pos = round(pos);
if pos < 1 || pos > n, pos = 1; end

geo = checkGeo(geo, angles);            % expands every field to per-view
ln = min(geo.sVoxel) / 2;               % coordinate-arrow length

%% Per-view rotations (ZYZ Euler, as the projectors interpret angles)
Rs = zeros(3, 3, n);
for i = 1:n
    Rs(:, :, i) = tiltedAxisRotation('z', angles(1, i)) * ...
                  tiltedAxisRotation('y', angles(2, i)) * ...
                  tiltedAxisRotation('z', angles(3, i));
end

%% Source / detector trajectories (lab x = beam, y = u, z = v)
scent = [geo.DSO; zeros(2, n)];                                  % 3xN
dcent = [-geo.DSD + geo.DSO; geo.offDetector(1, :); geo.offDetector(2, :)];
stj = zeros(3, n); dtj = zeros(3, n);
if strcmp(rotation, 'SD')
    for i = 1:n
        stj(:, i) = Rs(:, :, i) * scent(:, i);
        dtj(:, i) = Rs(:, :, i) * dcent(:, i);
    end
else
    stj(:, pos) = Rs(:, :, pos) * scent(:, pos);
    dtj(:, pos) = Rs(:, :, pos) * dcent(:, pos);
end
stj(2, :) = stj(2, :) + geo.COR;        % COR shifts source AND detector in y
dtj(2, :) = dtj(2, :) + geo.COR;

%% Detector orientation and cube
ddp = 30;                               % drawn detector depth (mm)
Rdet = zeros(3, 3, n);
for i = 1:n
    r = geo.rotDetector(:, i);
    Rdet(:, :, i) = Rs(:, :, i) * tiltedAxisRotation('x', r(1)) * ...
                    tiltedAxisRotation('y', r(2)) * tiltedAxisRotation('z', r(3));
end
dsz = [ddp; geo.sDetector(1); geo.sDetector(2)];     % [beam; u; v]
dverts = cubeVertices(dtj, dsz, Rdet, [-ddp/2; 0; 0]);

%% Object cube (offOrigin is [x;y;z] in MATLAB)
otj = geo.offOrigin;
otj(2, :) = otj(2, :) + geo.COR;
if strcmp(rotation, 'obj')
    Robj = permute(Rs, [2, 1, 3]);      % object rotates the other way
else
    Robj = repmat(eye(3), [1, 1, n]);
end
overts = cubeVertices(otj, geo.sVoxel, Robj, [0; 0; 0]);
FACES = [2 3 7 6; 1 2 3 4; 5 6 7 8; 1 2 6 5; 3 4 8 7; 1 4 8 5];

%% Figure
h = figure('Name', 'Cone Beam Computed Tomography geometry', 'Color', 'w', ...
           'Position', [100, 100, 900, 720]);
ax = axes('Parent', h); hold(ax, 'on'); grid(ax, 'on');
xlabel(ax, 'X'); ylabel(ax, 'Y'); zlabel(ax, 'Z');
view(ax, 52, 26);

plot3(ax, stj(1, :), stj(2, :), stj(3, :), '.', 'Color', [0.87 0.63 0.87], 'MarkerSize', 6);
plot3(ax, dtj(1, :), dtj(2, :), dtj(3, :), '.', 'Color', [0.53 0.81 0.92], 'MarkerSize', 6);
plot3(ax, otj(1, :), otj(2, :), otj(3, :), '.', 'Color', [0.5 0.5 0.5], 'MarkerSize', 6);

hS = plot3(ax, stj(1, pos), stj(2, pos), stj(3, pos), 'o', 'MarkerSize', 6, ...
           'MarkerFaceColor', [0.87 0.63 0.87], 'Color', [0.87 0.63 0.87]);
hD = plot3(ax, dtj(1, pos), dtj(2, pos), dtj(3, pos), 'o', 'MarkerSize', 6, ...
           'MarkerFaceColor', [0.53 0.81 0.92], 'Color', [0.53 0.81 0.92]);
tS = text(ax, stj(1, pos), stj(2, pos), stj(3, pos) + 30, 'S');
tD = text(ax, dtj(1, pos), dtj(2, pos), dtj(3, pos) + 30, 'D');

hDet = patch(ax, 'Vertices', dverts(:, :, pos), 'Faces', FACES, ...
             'FaceColor', [0.65 0.16 0.16], 'FaceAlpha', 0.2, 'EdgeColor', 'k');
hObj = patch(ax, 'Vertices', overts(:, :, pos), 'Faces', FACES, ...
             'FaceColor', 'c', 'FaceAlpha', 0.2, 'EdgeColor', 'k');

quiver3(ax, 0, 0, 0, ln, 0, 0, 0, 'r', 'LineWidth', 1.5);
quiver3(ax, 0, 0, 0, 0, ln, 0, 0, 'b', 'LineWidth', 1.5);
quiver3(ax, 0, 0, 0, 0, 0, ln, 0, 'g', 'LineWidth', 1.5);
text(ax, -10, -10, -10, 'O');

% beam: central ray + the four rays to the source-facing detector corners
% (cube face 1 = the -beam face, vertices 2 3 7 6)
front = FACES(1, :);
hBeam = gobjects(4, 1);
for k = 1:4
    c = dverts(front(k), :, pos);
    hBeam(k) = plot3(ax, [stj(1, pos), c(1)], [stj(2, pos), c(2)], [stj(3, pos), c(3)], 'y');
end
hC = plot3(ax, [stj(1, pos), dtj(1, pos)], [stj(2, pos), dtj(2, pos)], ...
           [stj(3, pos), dtj(3, pos)], 'Color', [1 0.6 0.8]);

axis(ax, 'equal');
hT = title(ax, frameTitle(pos));

%% Animate
if ~animate
    return
end
frames = 1:n;
gif = '';
if ~isempty(fname)
    gif = [fname, '_geometry.gif'];
end
for k = frames
    set(hObj, 'Vertices', overts(:, :, k));
    if strcmp(rotation, 'SD')
        set(hS, 'XData', stj(1, k), 'YData', stj(2, k), 'ZData', stj(3, k));
        set(hD, 'XData', dtj(1, k), 'YData', dtj(2, k), 'ZData', dtj(3, k));
        set(tS, 'Position', [stj(1, k), stj(2, k), stj(3, k) + 30]);
        set(tD, 'Position', [dtj(1, k), dtj(2, k), dtj(3, k) + 30]);
        set(hDet, 'Vertices', dverts(:, :, k));
        set(hC, 'XData', [stj(1, k), dtj(1, k)], 'YData', [stj(2, k), dtj(2, k)], ...
                'ZData', [stj(3, k), dtj(3, k)]);
        for j = 1:4
            c = dverts(front(j), :, k);
            set(hBeam(j), 'XData', [stj(1, k), c(1)], 'YData', [stj(2, k), c(2)], ...
                          'ZData', [stj(3, k), c(3)]);
        end
    end
    set(hT, 'String', frameTitle(k));
    drawnow;
    if ~isempty(gif)
        fr = getframe(h);
        [im, map] = rgb2ind(frame2im(fr), 256);
        if k == frames(1)
            imwrite(im, map, gif, 'gif', 'LoopCount', inf, 'DelayTime', 1/30);
        else
            imwrite(im, map, gif, 'gif', 'WriteMode', 'append', 'DelayTime', 1/30);
        end
    end
end

%% ---------------------------------------------------------------------
function s = frameTitle(k)
    rd = geo.rotDetector(:, k) * 180 / pi;
    if strcmp(rotation, 'SD')
        ang = angles(:, k) * 180 / pi;
        who = 'Source';
    else
        ang = -angles(:, k) * 180 / pi;
        who = 'Object';
    end
    s = sprintf(['CBCT geometry in scale - %s at angle [%.1f, %.1f, %.1f]%c\n' ...
                 'Detector rotation [%.1f, %.1f, %.1f]%c   ' ...
                 'Detector offset [u %.1f, v %.1f] mm'], ...
                who, ang(1), ang(2), ang(3), char(176), ...
                rd(1), rd(2), rd(3), char(176), ...
                geo.offDetector(1, k), geo.offDetector(2, k));
end
end

%% ------------------------------------------------------------------------
function V = cubeVertices(centre, sz, R, offcent)
%CUBEVERTICES 8x3xN corner coordinates of cuboids centred at CENTRE (3xN),
%   of size SZ (3x1), rotated by R (3x3xN), shifted by OFFCENT (3x1) before
%   rotation. Corner order matches the FACES table in the caller.
CORNERS = [-1 -1 -1; 1 -1 -1; 1 1 -1; -1 1 -1; -1 -1 1; 1 -1 1; 1 1 1; -1 1 1];
n = size(centre, 2);
V = zeros(8, 3, n);
base = CORNERS .* (sz(:)' / 2) + offcent(:)';
for i = 1:n
    V(:, :, i) = base * R(:, :, i)' + centre(:, i)';
end
end
